Testing fails over VPN

Tests fail when connected via VPN.  The error is:

```
Caused by: java.net.BindException: Problem binding to r5060116.mayo.edu/172.24.108.254:0 : Can't assign requested address
```

This happens because the class `org.apache.hadoop.hbase.regionserver.RSRpcServices` tries to be smart about looking up the listen address.  Over VPN it gets something like `r5060116.mayo.edu`, but that can't be resolved to a proper IP, and the bind fails.

This code sets the value.

```
 public static String getHostname(Configuration conf, boolean isMaster)
      throws UnknownHostException {
    String hostname = conf.get(isMaster? HRegionServer.MASTER_HOSTNAME_KEY :
      HRegionServer.RS_HOSTNAME_KEY);
    if (hostname == null || hostname.isEmpty()) {
      String masterOrRS = isMaster ? "master" : "regionserver";
      return Strings.domainNamePointerToHostName(DNS.getDefaultHost(
        conf.get("hbase." + masterOrRS + ".dns.interface", "default"),
        conf.get("hbase." + masterOrRS + ".dns.nameserver", "default")));
    } else {
      LOG.info("hostname is configured to be " + hostname);
      return hostname;
    }
  }
```

Is this worth fixing?  Testing will never work properly without a setting `HRegionServer.MASTER_HOSTNAME_KEY` to a proper value.

--- a/README.md
+++ b/README.md
@@ -52,10 +52,10 @@ Should be mostly self-explanatory.
 To [build an `assembly`](http://maven.apache.org/plugins/maven-assembly-plugin/), use the `assembly:single` goal:
 
 ```bash
-mvn assembly:single
+mvn deploy assembly:single
 ```
 
-This will build `target/hadoop-commons-1.0-SNAPSHOT-jar-with-dependencies.jar`.
+This will build `target/hadoop-commons-1.0-SNAPSHOT-jar-with-dependencies.jar`.  It is **critical** that the command line be used as is!  `assembly:single` alone will [not build the correct executable jar](http://stackoverflow.com/a/574650/334619)!
 
 # Development
 

--- a/README.md
+++ b/README.md
@@ -52,7 +52,7 @@ Should be mostly self-explanatory.
 To [build an `assembly`](http://maven.apache.org/plugins/maven-assembly-plugin/), use the `assembly:single` goal:
 
 ```bash
-mvn deploy assembly:single
+mvn package assembly:single
 ```
 
 This will build `target/hadoop-commons-1.0-SNAPSHOT-jar-with-dependencies.jar`.  It is **critical** that the command line be used as is!  `assembly:single` alone will [not build the correct executable jar](http://stackoverflow.com/a/574650/334619)!

--- a/README.md
+++ b/README.md
@@ -26,6 +26,13 @@ To include the (longer) integration tests:
 mvn failsafe:integration-test
 ```
 
+Getting Maven to [fetch source and javadoc](http://tedwise.com/2010/01/27/maven-micro-tip-get-sources-and-javadocs/) is easy:
+
+```bash
+mvn dependency:sources
+mvn dependency:resolve -Dclassifier=javadoc
+```
+
 ### Dependancies
 
 To add a dependency, edit the `pom.xml` file, looking for the `<dependancies>` section, *e.g.*

--- a/README.md
+++ b/README.md
@@ -47,6 +47,16 @@ To add a dependency, edit the `pom.xml` file, looking for the `<dependancies>` s
 
 Should be mostly self-explanatory.
 
+## Deployment
+
+To [build an `assembly`](http://maven.apache.org/plugins/maven-assembly-plugin/), use the `assembly:single` goal:
+
+```bash
+mvn assembly:single
+```
+
+This will build `target/hadoop-commons-1.0-SNAPSHOT-jar-with-dependencies.jar`.
+
 # Development
 
 HadoopCommons uses [Hadoop mini-clusters](http://www.lopakalogic.com/articles/hadoop-articles/hadoop-testing-with-minicluster/) for local testing / debugging (see [https://github.com/sakserv/hadoop-mini-clusters](https://github.com/sakserv/hadoop-mini-clusters) for details).  At runtime, the `AutoConfigure` class provides a static method `getConfiguration()` to return a Hadoop configuration.  This configuration is either:

--- a/src/main/java/edu/mayo/hadoop/commons/hbase/AutoConfigure.java
+++ b/src/main/java/edu/mayo/hadoop/commons/hbase/AutoConfigure.java
@@ -1,23 +1,27 @@
 package edu.mayo.hadoop.commons.hbase;
 
-import edu.mayo.hadoop.commons.minicluster.MiniClusterUtil;
-import org.apache.hadoop.fs.Path;
-import org.apache.hadoop.hbase.HBaseConfiguration;
-import org.apache.log4j.Logger;
-
 import java.io.InputStream;
 import java.nio.file.Files;
 import java.nio.file.Paths;
 import java.util.Properties;
 
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.HBaseConfiguration;
+import org.apache.log4j.Logger;
+
+import edu.mayo.hadoop.commons.minicluster.MiniClusterUtil;
+
 /**
  * Find a HBase configuration based on simple heuristics.
  * <p>
- * This class will determine an HBase configuration to return.  A simple set of rules are as follows:
+ * This class will determine an HBase configuration to return. A simple set of
+ * rules are as follows:
  * <p>
  * <ul>
  * <li>if <code>/etc/hbase/conf</code> exists, use that configuration (HDP)</li>
- * <li>else start up a <a href="http://www.lopakalogic.com/articles/hadoop-articles/hadoop-testing-with-minicluster/">mini-cluster</a></li>
+ * <li>else start up a <a href=
+ * "http://www.lopakalogic.com/articles/hadoop-articles/hadoop-testing-with-minicluster/">
+ * mini-cluster</a></li>
  * </ul>
  * </p>
  * Created by Daniel Blezek on 2/23/16.
@@ -25,7 +29,7 @@ import java.util.Properties;
 public class AutoConfigure {
     static Logger logger = Logger.getLogger(AutoConfigure.class);
 
-    public static   org.apache.hadoop.conf.Configuration getConfiguration() throws Exception {
+    public static org.apache.hadoop.conf.Configuration getConfiguration() throws Exception {
 
         // Check for a site wide configuration for HDP
         String configPath = "/etc/hbase/conf/hbase-site.xml";
@@ -37,11 +41,12 @@ public class AutoConfigure {
         }
 
         // Return a mini-cluster
-        String HBASE_PROPS = "/hbase.properties";
-        Properties props;
+        Properties props = new Properties();
 
-        try (InputStream stream = AutoConfigure.class.getResourceAsStream(HBASE_PROPS)) {
-            props = MiniClusterUtil.loadPropertiesStream(stream);
+        for (String resourceName : new String[]{"/hbase.properties", "/hdfs.properties", "/yarn.properties", "/zookeeper.properties"}) {
+            try (InputStream stream = AutoConfigure.class.getResourceAsStream(resourceName)) {
+                props.putAll(MiniClusterUtil.loadPropertiesStream(stream));
+            }
         }
 
         MiniClusterUtil.startHBASE(props);

--- a/src/main/java/edu/mayo/hadoop/commons/hbase/HBaseConnector.java
+++ b/src/main/java/edu/mayo/hadoop/commons/hbase/HBaseConnector.java
@@ -1,15 +1,15 @@
 package edu.mayo.hadoop.commons.hbase;
 
+import java.io.IOException;
+
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.hbase.HBaseConfiguration;
 import org.apache.hadoop.hbase.client.Connection;
 import org.apache.hadoop.hbase.client.ConnectionFactory;
 
-import java.io.IOException;
-
 /**
- * Connects the code to HBase via the hbase client.
- * Created by m102417 on 1/26/16.
+ * Connects the code to HBase via the hbase client. Created by m102417 on
+ * 1/26/16.
  */
 public class HBaseConnector {
 

--- a/src/main/java/edu/mayo/hadoop/commons/minicluster/MiniClusterUtil.java
+++ b/src/main/java/edu/mayo/hadoop/commons/minicluster/MiniClusterUtil.java
@@ -6,6 +6,8 @@ import java.io.InputStream;
 import java.util.Properties;
 
 import org.apache.hadoop.conf.Configuration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import com.github.sakserv.minicluster.config.ConfigVars;
 import com.github.sakserv.minicluster.impl.HbaseLocalCluster;
@@ -25,6 +27,7 @@ import com.github.sakserv.minicluster.impl.ZookeeperLocalCluster;
  * multiple copies of these services because it is expensive!
  */
 public class MiniClusterUtil {
+    private static final Logger LOG = LoggerFactory.getLogger(MiniClusterUtil.class);
 
     // variables to track what services are currently started in the jvm env.
     private static boolean zookeeperStarted = false;
@@ -101,6 +104,13 @@ public class MiniClusterUtil {
                     .setHbaseRootDir(props.getProperty(ConfigVars.HBASE_ROOT_DIR_KEY)).setZookeeperPort(Integer.parseInt(props.getProperty(ConfigVars.ZOOKEEPER_PORT_KEY)))
                     .setZookeeperConnectionString(props.getProperty(ConfigVars.ZOOKEEPER_CONNECTION_STRING_KEY)).setZookeeperZnodeParent(props.getProperty(ConfigVars.HBASE_ZNODE_PARENT_KEY))
                     .setHbaseWalReplicationEnabled(Boolean.parseBoolean(props.getProperty(ConfigVars.HBASE_WAL_REPLICATION_ENABLED_KEY))).setHbaseConfiguration(new Configuration()).build();
+            hbaseLocalCluster.getHbaseConfiguration().set("hbase.master.hostname", props.getProperty("hbase.master.hostname"));
+
+            // Set all our properties
+            for (Object key : props.keySet()) {
+                LOG.info("Setting " + key.toString() + ": " + props.getProperty(key.toString()));
+                hbaseLocalCluster.getHbaseConfiguration().set(key.toString(), props.getProperty(key.toString()));
+            }
             hbaseLocalCluster.start();
             hbaseStarted = true;
         }

--- a/src/main/resources/hbase.properties
+++ b/src/main/resources/hbase.properties
@@ -1,19 +1,13 @@
-# Zookeeper
-zookeeper.temp.dir=embedded_zk
-zookeeper.host=127.0.0.1
-zookeeper.port=22010
-zookeeper.connection.string=127.0.0.1:22010
 
 # HBase
+
+# Don't lookup the hostname, specify localhost
+hbase.master.hostname=localhost
+hbase.regionserver.hostname=localhost
+
 hbase.master.port=25111
 hbase.master.info.port=-1
 hbase.num.region.servers=1
 hbase.root.dir=embedded_hbase
 hbase.znode.parent=/hbase-unsecure
 hbase.wal.replication.enabled=false
-
-# HBase Test
-hbase.test.table.name=hbase_test_table
-hbase.test.col.family.name=cf1
-hbase.test.col.qualifier.name=cq1
-hbase.test.num.rows.to.put=50

--- a/src/test/java/edu/mayo/hadoop/commons/examples/HBaseITCase.java
+++ b/src/test/java/edu/mayo/hadoop/commons/examples/HBaseITCase.java
@@ -5,8 +5,6 @@ import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.fail;
 
 import java.io.IOException;
-import java.io.InputStream;
-import java.util.Properties;
 
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.hbase.HColumnDescriptor;
@@ -24,14 +22,13 @@ import org.apache.hadoop.hbase.client.Result;
 import org.apache.hadoop.hbase.client.Table;
 import org.apache.hadoop.hbase.util.Bytes;
 import org.junit.AfterClass;
-import org.junit.BeforeClass;
 import org.junit.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.github.sakserv.minicluster.config.ConfigVars;
 import com.google.protobuf.ServiceException;
 
+import edu.mayo.hadoop.commons.hbase.AutoConfigure;
 import edu.mayo.hadoop.commons.hbase.HBaseUtil;
 import edu.mayo.hadoop.commons.minicluster.MiniClusterUtil;
 
@@ -39,22 +36,6 @@ public class HBaseITCase {
 
     // Logger
     private static final Logger LOG = LoggerFactory.getLogger(HBaseITCase.class);
-    // properties file
-    public static final String HBASE_PROPS = "/hbase.properties";
-    private static Properties props;
-
-    static {
-        try (InputStream stream = HBaseITCase.class.getResourceAsStream(HBASE_PROPS)) {
-            props = MiniClusterUtil.loadPropertiesStream(stream);
-        } catch (IOException e) {
-            LOG.error("Unable to load property file: {}", HBASE_PROPS);
-        }
-    }
-
-    @BeforeClass
-    public static void setUp() throws Exception {
-        MiniClusterUtil.startHBASE(props);
-    }
 
     @AfterClass
     public static void tearDown() throws Exception {
@@ -63,7 +44,7 @@ public class HBaseITCase {
 
     @Test
     public void running() throws Exception {
-        Configuration configuration = MiniClusterUtil.getHbaseLocalCluster().getHbaseConfiguration();
+        Configuration configuration = AutoConfigure.getConfiguration();
         try {
             HBaseAdmin.checkHBaseAvailable(configuration);
         } catch (MasterNotRunningException mnre) {
@@ -85,17 +66,17 @@ public class HBaseITCase {
     public void testHbaseLocalCluster() throws Exception {
 
         LOG.info("Establishing a connection with HBase");
-        Configuration configuration = MiniClusterUtil.getHbaseLocalCluster().getHbaseConfiguration();
+        Configuration configuration = AutoConfigure.getConfiguration();
         try (Connection hcon = ConnectionFactory.createConnection(configuration)) {
             HBaseUtil hutil = new HBaseUtil(hcon);
 
             LOG.info("Drop Tables in case things did not cleanup correctly in the past");
             hutil.dropAll();
 
-            String tableName = props.getProperty(ConfigVars.HBASE_TEST_TABLE_NAME_KEY);
-            String colFamName = props.getProperty(ConfigVars.HBASE_TEST_COL_FAMILY_NAME_KEY);
-            String colQualiferName = props.getProperty(ConfigVars.HBASE_TEST_COL_QUALIFIER_NAME_KEY);
-            Integer numRowsToPut = Integer.parseInt(props.getProperty(ConfigVars.HBASE_TEST_NUM_ROWS_TO_PUT_KEY));
+            String tableName = "hbase_test_table";
+            String colFamName = "cf1";
+            String colQualiferName = "ca1";
+            Integer numRowsToPut = 50;
 
             LOG.info("HBASE: Creating table {} with column family {}", tableName, colFamName);
             createHbaseTable(hcon, tableName, colFamName, configuration);

--- a/src/test/java/edu/mayo/hadoop/commons/examples/SparkHBaseITCase.java
+++ b/src/test/java/edu/mayo/hadoop/commons/examples/SparkHBaseITCase.java
@@ -1,69 +1,37 @@
 package edu.mayo.hadoop.commons.examples;
 
-import edu.mayo.hadoop.commons.minicluster.MiniClusterUtil;
-import it.nerdammer.spark.hbase.*;
+import java.util.ArrayList;
 
 import org.apache.spark.SparkConf;
-import org.apache.spark.api.java.JavaRDD;
 import org.apache.spark.api.java.JavaSparkContext;
-import org.apache.spark.rdd.RDD;
-import org.junit.BeforeClass;
 import org.junit.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.io.IOException;
-import java.io.InputStream;
-import java.util.ArrayList;
-import java.util.Properties;
-import java.util.stream.IntStream;
-
-import static java.util.stream.IntStream.*;
-
 /**
  * Created by m102417 on 2/15/16.
  *
- * Checks to make sure that Spark and HBase can work together using the HBase-Spark Connector:
+ * Checks to make sure that Spark and HBase can work together using the
+ * HBase-Spark Connector:
  *
  *
  */
 public class SparkHBaseITCase {
 
-
     // Logger
     private static final Logger LOG = LoggerFactory.getLogger(SparkHBaseITCase.class);
-    // properties file
-    public static final String HBASE_PROPS = "/hbase.properties";
-    private static Properties props;
-
-    static {
-        try (InputStream stream = HBaseITCase.class.getResourceAsStream(HBASE_PROPS)) {
-            props = MiniClusterUtil.loadPropertiesStream(stream);
-        } catch (IOException e) {
-            LOG.error("Unable to load property file: {}", HBASE_PROPS);
-        }
-    }
-
-    @BeforeClass
-    public static void setUp() throws Exception {
-        MiniClusterUtil.startHBASE(props);
-    }
-
     @Test
-    public void testConnect()throws Exception {
+    public void testConnect() throws Exception {
 
-        //make a connection on localhost to spark
+        // make a connection on localhost to spark
         SparkConf conf = new SparkConf().setMaster("local").setAppName("My App");
         JavaSparkContext sc = new JavaSparkContext(conf);
 
-        ArrayList values = new ArrayList();
-        for(int i=0;i<100;i++){values.add(i);}
+        ArrayList<Integer> values = new ArrayList<>();
+        for (int i = 0; i < 100; i++) {
+            values.add(i);
+        }
         // rdd = sc.parallelize(values).map(x -> x);
-
-
-
-
-
 
     }
 

